juju/juju
# Model controller HA

## Description of change

Commit 1 - state/model changes to introduce controller node entities
Commit 2 - upgrade steps to create controller nodes from existing machines

We create new controller node docs in the controllers collection; these represent machines in the mongo replicaset and their voting status. The idea is that we will remove "novote" and "hasvote" from the machine doc, and all associated HA watching and status etc will be drive off the controller nodes. This properly separates concerns and allows k8s deployments to avoid having to create machine docs.

This PR is the first of several - next steps are to move watchers across, refactor peer grouper worker, then delete the HA stuff off machine.

Several of our unit tests were broken (and so is our current HA modelling on machine). The tests set "wants vote" to true even though a machine is not a controller machine (ie it is missing the ManageMachines job). The new controller node modelling enforces correct behaviour so tests neede dto be fixed to account for that.

## QA steps

bootstrap lxd
use mongo client to check controllers collection that there's a single controller node
enable-ha
check that there's now 3 controller nodes and their contents
check juju show-controllers for correct info
remove a controller machine
repeat the above checks (2 nodes now)
add a new machine
enable-ha --to <id>
repeat the above checks (3 nodes again)

bootstrap 2.6
enable-ha
upgrade to this version
check the controller nodes have been created
